skimage/morphology/tests/test_extrema.py

```diff
@@ -171,6 +171,82 @@ def test_extrema_float(self):
         error = diff(expected_result, out)
         assert error < eps
 
+    def test_h_maxima_float_image(self):
+        """specific tests for h-maxima float image type"""
+        w = 10
+        x, y = np.mgrid[0:w, 0:w]
+        data = 20 - 0.2 * ((x - w / 2) ** 2 + (y - w / 2) ** 2)
+        data[2:4, 2:4] = 40
+        data[2:4, 7:9] = 60
+        data[7:9, 2:4] = 80
+        data[7:9, 7:9] = 100
+        data = data.astype(np.float32)
+
+        expected_result = np.zeros_like(data)
+        expected_result[(data > 19.9)] = 1.0
+
+        for h in [1.0e-12, 1.0e-6, 1.0e-3, 1.0e-2, 1.0e-1, 0.1]:
+            out = extrema.h_maxima(data, h)
+            error = diff(expected_result, out)
+            assert error < eps
+
+    def test_h_maxima_float_h(self):
+        """specific tests for h-maxima float h parameter"""
+        data = np.array([[0, 0, 0, 0, 0],
+                         [0, 3, 3, 3, 0],
+                         [0, 3, 4, 3, 0],
+                         [0, 3, 3, 3, 0],
+                         [0, 0, 0, 0, 0]], dtype=np.uint8)
+
+        h_vals = np.linspace(1.0, 2.0, 100)
+        failures = 0
+        for i in range(h_vals.size):
+            maxima = extrema.h_maxima(data, h_vals[i])
+
+            if (maxima[2, 2] == 0):
+                failures += 1
+
+        assert (failures == 0)
+
+    def test_h_maxima_large_h(self):
```

Comment:
Not sure, this tests fails for me with current master, but I might be doing something incorrectly in terms of switching back and forth between the branches.

Reply:
I'm using Python v3.6.9, numpy v1.18.1. Not sure what else might be relevant to differences in the test results.

Reply:
If I add 10 to all the values I can also make the integer version of this test fail. It was passing before because the `_subtract_constant_clip` function was setting all the values to 0.

```
        data = np.array([[10, 10, 10, 10, 10],
                         [10, 13, 13, 13, 10],
                         [10, 13, 14, 13, 10],
                         [10, 13, 13, 13, 10],
                         [10, 10, 10, 10, 10]], dtype=np.uint8)

        maxima = extrema.h_maxima(data, 5)
        assert (np.sum(maxima) == 0)
```

Reply:
As a reference I looked at the results from [this](https://imagej.net/Interactive_Watershed) ImageJ plugin on the above test image (saved as a tif). Using 'Hmaxima local maximum detection' and setting 'Threshold (minimum gray value)' to 0.0 and 'Minimum grey value distance between local optima' to 5.0 I get the result that I expected, which is that there are no local maxima greater than 5.0.

```
data = np.array([[10, 10, 10, 10, 10],
                 [10, 13, 13, 13, 10],
                 [10, 13, 14, 13, 10],
                 [10, 13, 13, 13, 10],
                 [10, 10, 10, 10, 10]], dtype=np.uint8)

tifffile.imsave("test.tif", data)
```

Reply:
If you agree I propose to fix this by first subtracting the minimum of `image` off `image`, then subtract `h` and finally set all values in `shifted_img` less than 0 to 0. This should work most of the time, but will introduce overflow issues, for example if image is `np.int` and we subtract -100 then any pixels in image that were greater than 28 will now be negative. So type upgrading may also be necessary.

Reply:
Hi @HazenBabcock thank you so much for your investigations. The type cast in a bit clumsy but still better than having a bug. Did you understand the origin of the bug though? I can't see how removing a value will solve this. Is it a problem of type overflow? Thanks again :-).

Reply:
I decided it would be better to add a check for a h value that is larger than the image range. In this case the function immediately returns an array of zeros. This way there is no need for type upgrading, but `h-maxima` is slower due to this additional check.

Reply:
works for me!